File: PopularShows/app/src/main/java/com/lgcode/popularshows/views/TvShowDetailView.java
package com.lgcode.popularshows.views;

import android.app.Activity;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.otto.Bus;
import com.squareup.picasso.Picasso;

import butterknife.BindView;
import butterknife.ButterKnife;
import glb.lg.popularshows.R;
import com.lgcode.popularshows.adapters.SimilarTVShowsAdapter;

/**
 * Created by leojg on 11/5/16.
 */
public class TvShowDetailView extends ActivityView {

    private Bus bus;

    @BindView(R.id.tv_show_detail_collapsing_toolbar) CollapsingToolbarLayout titleToolbar;
    @BindView(R.id.tv_show_overview) TextView tvShowOverview;
    @BindView(R.id.similar_tv_shows_list) RecyclerView similarTVShowsList;
    @BindView(R.id.tv_show_hero_image) ImageView tvShowHeroImage;

    private boolean isLoading;
    private boolean isLastPage;
    private int pageSize;


    public TvShowDetailView(AppCompatActivity activity, final Bus bus) {
        super(activity);
        this.bus = bus;

        setHomeAsUpEnabled(true);
        ButterKnife.bind(this, activity);
        similarTVShowsList.setLayoutManager(new LinearLayoutManager(getActivity(), LinearLayoutManager.HORIZONTAL, false));
        similarTVShowsList.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);

                LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();

                int visibleItemCount = layoutManager.getChildCount();
                int totalItemCount = layoutManager.getItemCount();
                int firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();

                if (!isLoading && !isLastPage) {
                    if ((visibleItemCount + firstVisibleItemPosition) >= totalItemCount
                            && firstVisibleItemPosition >= 0
                            && totalItemCount >= pageSize) {
                        bus.post(new SimilarTVShowsNewPageRequestEvent());
                        //TODO: Send event to request more items
                    }
                }
            }
        });


    }

    public void setAdapter(SimilarTVShowsAdapter adapter) {
        similarTVShowsList.setAdapter(adapter);
    }

    public void updateSimilarTVShowList(Integer currentPageSize) {
        similarTVShowsList.getAdapter().notifyDataSetChanged();
        pageSize = currentPageSize;
    }

    public void updateTVShow(String title, String overview, String heroImage) {
        Activity activity = this.getActivity();
        if (activity == null) {
            return;
        }
        titleToolbar.setTitle(title);
        tvShowOverview.setText(overview);
        Picasso.with(activity).load(heroImage).into(tvShowHeroImage);
    }

    //Events
    public class SimilarTVShowsNewPageRequestEvent {
    }
}

File: KtPopularShows/app/src/main/java/example/lgcode/ktpopularshows/mvp/view/TVShowDetailView.kt
package example.lgcode.ktpopularshows.mvp.view

import android.support.v7.app.ActionBar
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import com.squareup.picasso.Picasso
import example.lgcode.ktpopularshows.BuildConfig
import example.lgcode.ktpopularshows.adapter.TVShowsAdapter
import example.lgcode.ktpopularshows.domain.TVShow
import kotlinx.android.synthetic.main.activity_detail_view.*
import java.lang.ref.WeakReference

class TVShowDetailView(val activity: AppCompatActivity): ActivityView(WeakReference(activity)) {

    init {
        if (activity?.tvshow_detail_toolbar != null) {
            activity.setSupportActionBar(activity.tvshow_detail_toolbar)
        }
        setHomeAsUpEnabled(true)
        activity.similar_tv_shows_list.layoutManager = LinearLayoutManager(activity, LinearLayoutManager.HORIZONTAL, false)

    }

    fun getListLinearLayoutManager(): LinearLayoutManager {
        return activity.similar_tv_shows_list.layoutManager as LinearLayoutManager
    }

    fun setShowListScrollListener(scrollListener: RecyclerView.OnScrollListener) {
        activity.similar_tv_shows_list.addOnScrollListener(scrollListener)
    }

    fun setAdapter(adapter: TVShowsAdapter) {
        activity.similar_tv_shows_list.adapter = adapter
    }

    fun updateTVShow(title: String, overview: String, heroImage: String) {
        if (activity == null) {
            return
        }

        activity.tv_show_detail_collapsing_toolbar.title = title
        activity.tv_show_overview.text = overview
        Picasso.with(activity)
                .load(StringBuilder(BuildConfig.HERO_IMAGE_BASE_URL).append(heroImage).toString())
                .into(activity.tv_show_hero_image)

    }

    fun updateList(tvShows: List<TVShow>) {
        (activity.similar_tv_shows_list.adapter as TVShowsAdapter).addAll(tvShows)
        activity.similar_tv_shows_list.adapter.notifyDataSetChanged()
    }

    private fun setHomeAsUpEnabled(showHomeAsUp: Boolean) {
        val actionBar = getActionBar()

        if (activity.tvshow_detail_toolbar == null || actionBar == null) {
            return
        }

        activity.tvshow_detail_toolbar.setNavigationOnClickListener({
            activity.onBackPressed()
        })

        actionBar.setDisplayHomeAsUpEnabled(showHomeAsUp)

    }

    private fun getActionBar(): ActionBar? {
        return activity.supportActionBar
    }

}